### ISSUE TITLE
Different inlining behavior with and without caches

### ISSUE THREAD

**Issue by aseemr** [FStarLang/FStar#1524]
`BufferOps.blit` is marked `inline_for_extraction`. For the following program:

```
module Test

open LowStar.Buffer
open LowStar.BufferOps
open FStar.HyperStack.ST

let foo (b1 b2:buffer int) :ST unit (requires (fun h0 -> live h0 b1 /\ live h0 b2 /\ length b1 == 10 /\ length b2 == 10 /\ disjoint b1 b2)) (ensures  (fun h0 _ h1 -> True))
  = LowStar.BufferOps.blit b1 0ul b2 0ul 2ul
```

If we extract it using `fstar.exe --codegen OCaml Test.fst`, the function gets properly inlined if there are no `ulib/` cache present. However when the cache is present, it is not inlined.

**Comment by mtzguido** [comment]
I'm seeing similar things, which affect verification in a branch I have. Here's a tiny repro:
```
module A
let _ = assert False (* Just to trigger a query *)
```
I can see a difference in the SMT context with cache on/off, which leads me to believe that some definitions do not make it into the cache.
```
$ ./bin/fstar.exe --log_queries A.fst
A.fst(3,8-3,14): (Error 19) assertion failed [....]
$ mv queries-A.smt2 q1.smt2
$ ./bin/fstar.exe --log_queries A.fst --cache_off
A.fst(3,8-3,14): (Error 19) assertion failed [....]
$ mv queries-A.smt2 q2.smt2
$ diff q1.smt2 q2.smt2
```
and I see, among the noise:
```diff
@@ -12184,6 +12184,19 @@ Prims.unit))
 :qid typing_Prims._assert))
 
 :named typing_Prims._assert))
+;;;;;;;;;;;;;;;;Equation for Prims._assert
+;;; Fact-ids: Name Prims._assert; Namespace Prims
+(assert (! 
+;; def=/home/guido/r/fstar/ulib/prims.fst(395,4-395,11); use=/home/guido/r/fstar/ulib/prims.fst(395,4-395,11)
+(forall ((@x0 Term))
+ (! (= (Prims._assert @x0)
+Tm_unit)
+ 
+
+:pattern ((Prims._assert @x0))
+:qid equation_Prims._assert))
+
+:named equation_Prims._assert))
 
 ; </end encoding Prims._assert>
 
@@ -12194,7 +12207,7 @@ Prims.unit))
```
In this case, I see this for
```
Prims._assert
Prims.assert_spinoff
Prims.cut
FStar.Pervasives.allow_inversion
FStar.Pervasives.ignore
FStar.Pervasives.dm4f_bind_range
```
note `FStar.Pervasives.ignore`! Our tactics often got stuck at this thing, so this might be the cause.

Still looking into it.

**Comment by aseemr** [comment]
I have a fix for the original issue, it had to do with extracted interfaces and inline for extraction flag. Will try to look into your example, may be it's related.

**Comment by mtzguido** [comment]
Thanks! But if it's not the same issue don't sweat it too much, it's not a big roadblock right now.

**Comment by mtzguido** [comment]
After Aseem quickly pointed me to the root of the issue, my problem is fixed in commit e25c4cc317f7b255ac85709eda317bca68e6444a.

**Comment by aseemr** [comment]
Thinking more about it, I think we have the right behavior currently. Extracted interfaces should not keep the definitions of effectful functions marked `inline_for_extraction` -- cross module inlining is the better way to solve that. If we start keeping these definitions, we end up increasing the verification context unnecessarily.

Closing this issue.
